Refactor and docs

vladimyr opened the pull request:
This PR brings in backward compatible public API changes to support passing optional `s3opts` param. JSDoc annotations are being added to support documentation generation. 🎉

vladimyr commented on index.js:
```diff
@@ -28,40 +40,22 @@ S3BlobStore.prototype.createReadStream = function (opts) {
   return stream;
 };
 
-S3BlobStore.prototype.uploadParams = function (opts) {
-  opts = Object.assign({}, opts, {
-    params: Object.assign({}, opts.params)
-  });
-
-  var filename = opts.name || opts.filename;
-  var key = opts.key || filename;
-  var contentType = opts.contentType;
-
-  var params = opts.params;
-  params.Bucket = params.Bucket || this.bucket;
-  params.Key = params.Key || key;
-
-  if (!contentType) {
-    contentType = filename ? mime.lookup(filename) : mime.lookup(opts.key);
-  }
-  if (contentType) params.ContentType = contentType;
-
-  return params;
-};
-
-S3BlobStore.prototype.downloadParams = function (opts) {
-  var params = this.uploadParams(opts);
-  delete params.ContentType;
-  return params;
-};
```
`store#uploadParams` & `store#downloadParams` have been removed and replaced with private `store#_s3params` helper method which shouldn't be a breaking change as they never were considered part of public API.

jb55 replied:
unless the behavior is exactly the same I would consider it a breaking change, so we can push this as v5 if so

vladimyr commented:
Ok I'll bring them back and mark as deprecated with warning saying they'll
be removed in upcoming v5 major version.

On Sun, Feb 16, 2020, 01:44 William Casarin <notifications@github.com>
wrote:

> *@jb55* commented on this pull request.
> ------------------------------
>
> In index.js
> <https://github.com/jb55/s3-blob-store/pull/23#discussion_r379866278>:
>
> > -S3BlobStore.prototype.uploadParams = function (opts) {
> -  opts = Object.assign({}, opts, {
> -    params: Object.assign({}, opts.params)
> -  });
> -
> -  var filename = opts.name || opts.filename;
> -  var key = opts.key || filename;
> -  var contentType = opts.contentType;
> -
> -  var params = opts.params;
> -  params.Bucket = params.Bucket || this.bucket;
> -  params.Key = params.Key || key;
> -
> -  if (!contentType) {
> -    contentType = filename ? mime.lookup(filename) : mime.lookup(opts.key);
> -  }
> -  if (contentType) params.ContentType = contentType;
> -
> -  return params;
> -};
> -
> -S3BlobStore.prototype.downloadParams = function (opts) {
> -  var params = this.uploadParams(opts);
> -  delete params.ContentType;
> -  return params;
> -};
>
> unless the behavior is exactly the same I would consider it a breaking
> change, so we can push this as v5 if so
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jb55/s3-blob-store/pull/23?email_source=notifications&email_token=AAI5YCAWX3BNVNDNS4FJYFLRDCD7HA5CNFSM4KV3TKZKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCVVZXLQ#discussion_r379866278>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAI5YCBDBQT2HYOOALC6N53RDCD7HANCNFSM4KV3TKZA>
> .
>

jb55 commented:
can we have the docs separate from the refactor? I would like to pull the docs for now and perhaps the refactor later after I have more time to review.

Commit wise I'm thinking doc commits first and then refactor commits after (with doc adjustments if needed)

jb55 commented:
needs rebase

vladimyr committed:
Refactor internals & update API docs

- extending all public API methods to support passing additional
  s3 parameters either through `opts` argument or dedicated
  `s3opts` argument
- introducing `store#_s3paramsi(opts, s3opts)` helper used for
  extracting additional s3 parameters
- deprecating undocumented `uploadParams` and `downloadParams`
  methods to prepare for new major release
- updating API documentation

vladimyr commented:
> needs rebase

✅ Done

jb55 commented:
>> needs rebase
>
> ✅ Done

awesome, much easier to see the changes now. thanks! will take a look at
it soon
